### PR TITLE
Voice recording UI state in ViewModel

### PR DESCRIPTION
**Pull request by ouchadam**
Inverts the `VoiceMessageRecorderView` update flow so that `TextComposerViewModel` controls the `RecordingUiState`

- Allows the dragging to cancel or lock mid way, the current behaviour forces whichever action the user starts dragging towards
- Fixes a small UI quirk where the recording layout would jump, caused by aligning to the mic button rather than the send button
- Fixes the dragging becoming inverted if dragging past the initial touch position
- Tentatively fixes https://github.com/vector-im/element-android/issues/4400 (I've not been able to reproduce on this branch at least)

Next step is to port the Voice draft to the SDK #4237 

| CANCEL | LOCK SEND | LOCK DELETE |
| --- | --- | --- |
after-voice-cancel|after-lock-send|after-voice-lock-delete

| CANCEL OR LOCK | HOLDING |
| --- | --- |
after-cancel-or-lock|after-voice-hold

### CODE REVIEW

**Comment by DoM1niC** [comment]
Nice One :) can you try to fix the multiple send issue if you send more than one Voice Message the following Messages will Stuck and can't be resended.

**Comment by ouchadam** [comment]
@DoM1niC I'll have a look since I'm in the area!

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/TextComposerViewState.kt [hunk 1]
```diff
@@ -44,13 +45,29 @@ sealed class SendMode(open val text: String) {
...
         val roomId: String,
         val canSendMessage: Boolean = true,
-        val isVoiceRecording: Boolean = false,
         val isSendButtonVisible: Boolean = false,
-        val sendMode: SendMode = SendMode.REGULAR("", false)
+        val sendMode: SendMode = SendMode.REGULAR("", false),
+        val voiceRecordingUiState: VoiceMessageRecorderView.RecordingUiState = VoiceMessageRecorderView.RecordingUiState.None
 ) : MavericksState {
 
+    val isVoiceRecording = when (voiceRecordingUiState) {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/home/room/detail/composer/TextComposerViewState.kt:53]: the `isRecording` and `isIdle` states are inferred from the UI state

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageViews.kt [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+import im.vector.app.R
+import im.vector.app.core.extensions.setAttributeBackground
+import im.vector.app.core.extensions.setAttributeTintedBackground
+import im.vector.app.core.extensions.setAttributeTintedImageResource
+import im.vector.app.core.utils.DimensionConverter
+import im.vector.app.databinding.ViewVoiceMessageRecorderBinding
+import im.vector.app.features.home.room.detail.composer.voice.VoiceMessageRecorderView.RecordingUiState
+import im.vector.app.features.home.room.detail.timeline.helper.VoiceMessagePlaybackTracker
+
+class VoiceMessageViews(
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageViews.kt:38]: all of the android widget interactions have been extracted into this delegate, to help keep the `VoiceMessageRecorderView` as simple as possible

File: changelog.d/4515.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Voice recording mic button refactor with small animation tweaks in preparation for voice drafts
```
Comment (ouchadam) [inline changelog.d/4515.misc:1]: wasn't 100% sure if the change log entry was needed as this is mostly refactoring (same as  https://github.com/vector-im/element-android/pull/4523)

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageRecorderView.kt [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+    override fun onVisibilityChanged(changedView: View, visibility: Int) {
+        super.onVisibilityChanged(changedView, visibility)
+        // onVisibilityChanged is called by constructor on api 21 and 22.
+        if (!this::voiceMessageViews.isInitialized) return
+        val parentChanged = changedView == this
+        voiceMessageViews.renderVisibilityChanged(parentChanged, visibility)
+    }
+
+    fun display(recordingState: RecordingUiState) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageRecorderView.kt:106]: Nit: in the codebase those methods are named `render()`

File: changelog.d/4515.misc (hunk 3 above)
Comment (bmarty) [reply changelog.d/4515.misc:1]: Always OK to have a changelog, if something goes wrong (I hope no!), we keep quickly see in the change history what have been done.

**Review by bmarty — APPROVED** [review]
LGTM, small remark on naming.
Did not test, will rather test after change on draft handling 👍

**Comment by bmarty** [comment]
Stop the CI, due to add changelog entry. Previous commit was green (sort of)

File: vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageRecorderView.kt (hunk 4 above)
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/features/home/room/detail/composer/voice/VoiceMessageRecorderView.kt:106]: ah, will update in the other PR

**Pull request merged**